server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -410,6 +441,204 @@ public void clusterChanged(ClusterChangedEvent event) {
+    ) {
+        return new PollClusterFormationStateTask(node, nodeToClusterFormationStateMap).pollUntilCancelled();
+    }
+
+    /*
+     * This inner class wraps the logic of polling a master-eligible node for its cluster formation information (which is needed in the
+     * event that the cluster cannot elect a master node).
+     */
+    // Non-private for testing
+    class PollClusterFormationStateTask {
+        /**
+         * The node that is being polled
+         */
+        private final DiscoveryNode node;
+        /**
+         * This is a reference to the global nodeToClusterFormationStateMap that was current at the time this object was constructed. The
+         * global map is recreated whenever the task is cancelled. Having this reference prevents accidental writes to that map after
+         * cancellation.
+         */
+        private final ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap;
+        /**
+         * This is a wrapper Cancellable. After polling begins, every time a new remote request is scheduled (about once every 10
+         * seconds) we get a new Cancellable. This wraps all of them so that we only have to cancel the single Cancellable that is
+         * initially returned from pollUntilCancelled() in order to cancel them all.
+         */
+        private final MultipleCancellablesWrapper multipleCancellablesWrapper;
+
+        /**
+         * This constructor is used to create the root task. It initializes the MultipleCancellablesWrapper that is shared between all
+         * the related tasks.
+         *
+         * @param node The node to poll for cluster formation information
+         * @param nodeToClusterFormationStateMap A reference to the global nodeToClusterFormationStateMap
+         */
+        PollClusterFormationStateTask(
+            DiscoveryNode node,
+            final ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap
+        ) {
+            this(node, nodeToClusterFormationStateMap, new MultipleCancellablesWrapper());
+        }
+
+        private PollClusterFormationStateTask(
+            DiscoveryNode node,
+            final ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap,
+            MultipleCancellablesWrapper multipleCancellablesWrapper
+        ) {
+            this.node = node;
+            this.nodeToClusterFormationStateMap = nodeToClusterFormationStateMap;
+            this.multipleCancellablesWrapper = multipleCancellablesWrapper;
+        }
+
+        /**
+         * This method returns a Cancellable quickly, but in the background schedules to query the remote node's cluster formation state
+         * in 10 seconds, and repeats doing that until cancel() is called on the returned Cancellable.
+         *
+         * @return
+         */
+        public Scheduler.Cancellable pollUntilCancelled() {
```

Comment:
Oh good point about PollClusterFormationStateTask's existence. In an earlier incarnation this was an actual Runnable, and `pollUntilCancelled()` was just `run()`. I improved that, but forgot to move the state out of `PollClusterFormationStateTask` and then get rid of it altogether. I think the signature will have to be a little different though. More on that in an upcoming comment.

Reply:
Here's the signature I have now:
`void fetchClusterFormationInfo(DiscoveryNode node, ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap, List<Scheduler.Cancellable> cancellables)`
I can't just return a single Cancellable because one call to the method can generate lots of Cancellables (since the method schedules calls of itself). So I'm passing in the array of Cancellables that the `cancelPollingClusterFormationInfo` acts on. This avoids having to have the `MultipleCancellablesWrapper` but it effectively does the same thing. And I passed in a nodeToClusterFormationStateMap because I can't have just a ClusterFormationInfoAction.Response Consumer because one possible outcome is an Exception rather than a Response. I'll see if making that a Consumer<ClusterFormationStateOrException> makes it more or less readable.

Reply:
OK the method now accepts a `Consumer<ClusterFormationStateOrException>` instead of a `ClusterFormationStateOrException> nodeToClusterFormationStateMap`. I think it helps out a little bit. The functionality is unchanged.

Reply:
I think this is a step in the right direction, thanks for implementing it ! 

>I can't just return a single Cancellable because one call to the method can generate lots of Cancellables (since the method schedules calls of itself)

Would this be an avenue for simplification?

Say the method signature is as follows
```
Scheduler.Cancellable fetchClusterFormationInfo(DiscoveryNode node, Consumer<ClusterFormationInfoAction.Response> responseConsumer)
```
This means the caller is responsible for re-calling `fetchClusterFormationInfo` method when the `responseConsumer` is called. 

I think this would remove the need for the `runAfter` call in that `runAfter(runBefore` sequence because it seems we'll want to reschedule irrespective of being able to get the response or not - so when `responseConsumer` is called we should be rescheduling. 
ie. we always call `responseConsumer` irrespective of how `fetchClusterInfoListener` completes:
```
                new ActionListenerResponseHandler<>(
                    ActionListener.runBefore(fetchClusterInfoListener, () -> Releasables.close(releasable)),
                    ClusterFormationInfoAction.Response::new
                )
```

Consuming the `fetchClusterFormationInfo` would look something along the lines of:

```   
        getMasterEligibleNodes().forEach(masterNode -> {
            Consumer<ClusterFormationStateOrException> responseConsumer = response -> {
                if (clusterFormationResponses != null) {
                    clusterFormationResponses.put(masterNode, response);
                }
            };
            
            Scheduler.ScheduledCancellable scheduleFetch = fetchClusterFormationInfo(masterNode, responseConsumer.andThen(response -> {
                if (clusterFormationInfoTasks != null) {
                    // reschedule the fetch if it wasn't cancelled already
                    clusterFormationInfoTasks.add(fetchClusterFormationInfo(masterNode, responseConsumer));
                }
            }));
            
            clusterFormationInfoTasks.add(scheduleFetch);
        });

```

This would reduce the scope of `fetchClusterFormationInfo` and the need for a `Consumer<Cancellable>`.

What do you think?

Reply:
Unless I'm misunderstanding something, we can't use a `Consumer<ClusterFormationInfoAction.Response>` here because then we'd have no way of handling exceptions, right? And we need to keep track of exceptions we encounter.
And even using a `Consumer<ClusterFormationStateOrException>` I don't think we can take the approach you're suggesting will work as-is because it misses retries on a `connectToNode` exception (https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR529). I could probably add in a separate consumer for `connectToNode`, but at that point it's starting to get even more complex. Is there any use case that is missed by the current code?

Reply:
Also I think your proposal exposes us to the race condition here right (although I'm not 100% sure that causes big problems)? https://github.com/elastic/elasticsearch/pull/88397#discussion_r919348039
But that is fixable.

Reply:
>And even using a Consumer<ClusterFormationStateOrException> I don't think we can take the approach you're suggesting will work as-is because it misses retries on a connectToNode exception (https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR529)

Ah apologies I did mean `Consumer<ClusterFormationStateOrException>`. But if the `connectToNode` raises an exception the response consumer will still be called https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR524 (and the responseConsumer will track the exception and reschedule the `fetch`). Am I misisng something? 

> Is there any use case that is missed by the current code?

I think it's a bit difficult to follow, that's why I suggested reducing the scope of `fetch...` and have a clear view over which callback reschedules the polling and when that happens